### PR TITLE
Replace snapshots with iterations throughout orchestra

### PR DESCRIPTION
**Pull request by jrbotros**

### CODE REVIEW

File: docs/source/api.rst [hunk 1]
```diff
@@ -70,27 +70,23 @@ All requests must be signed using `HTTP signatures <http://tools.ietf.org/html/d
               "sample_step_slug": {
                   "id": 456,
                   "project": 123,
-                  "status": "Processing",
+                  "status": "PENDING_REVIEW",
```
Comment (marcua) [inline docs/source/api.rst]: does it make sense that this changed from pretty print to enum value?

File: docs/source/api.rst [hunk 2]
```diff
@@ -70,27 +70,23 @@ All requests must be signed using `HTTP signatures <http://tools.ietf.org/html/d
...
-                                      "work_time_seconds": 3660,
-                                      "datetime": "2015-09-23T20:16:15.821171",
-                                      "data": {
-                                          "sample_data_item": "sample_data_value_old",
-                                          "__version": 1
-                                      },
-                                      "type": 0
-                                  }
-                              ]
+                          "iterations": {
```
Comment (marcua) [inline docs/source/api.rst]: `iterations` suggests a list, but there's a dict

File: orchestra/analytics/latency.py [hunk 3]
```diff
@@ -1,202 +0,0 @@
-import itertools
```
Comment (marcua) [inline orchestra/analytics/latency.py:1]: nbd just deleting some old files here nbd

File: docs/source/api.rst (hunk 1 above)
Comment (jrbotros) [inline docs/source/api.rst]: Don't think this actually changed, just messed up the docs for whatever reason

File: docs/source/api.rst (hunk 2 above)
Comment (jrbotros) [inline docs/source/api.rst]: Oops, it is a list—will update docs

File: orchestra/machine_tasks.py [hunk 4]
```diff
@@ -88,14 +86,8 @@ def execute(project_id, step_slug):
...
         task.save()
-        task_assignment.snapshots['snapshots'].append(
-            {'data': task_assignment.in_progress_task_data,
-             'datetime': timezone.now().isoformat(),
-             'type': TaskAssignment.SnapshotType.SUBMIT,
-             'work_time_seconds': 0})
-        task_assignment.save()
 
-        iteration = get_latest_iteration(task_assignment)
+        iteration = get_latest_iteration(task_assignment.id)
```
Comment (marcua) [inline orchestra/machine_tasks.py]: is this just some api cleanup for consistency?

File: docs/source/api.rst [hunk 5]
```diff
@@ -70,27 +70,23 @@ All requests must be signed using `HTTP signatures <http://tools.ietf.org/html/d
...
+                              "id": 92134,
+                              "start_datetime": "2015-09-20T12:02:14.214553",
+                              "end_datetime": "2015-09-23T20:16:15.821171",
+                              "data": {
```
Comment (marcua) [inline docs/source/api.rst]: `submitted_data`

File: orchestra/interface_api/project_management/views.py [hunk 6]
```diff
@@ -65,14 +63,9 @@ def reassign_assignment_api(request):
...
     try:
-        revert_datetime = datetime.utcfromtimestamp(body['revert_datetime'])
-        revert_datetime = pytz.utc.localize(revert_datetime)
-    except:
-        raise BadRequest('Invalid timestamp provided.')
-
-    try:
-        audit = revert_task_to_datetime(
-            body['task_id'], revert_datetime, body['fake'])
+        audit = revert_task_to_iteration(
```
Comment (marcua) [inline orchestra/interface_api/project_management/views.py]: from @jrbotros: "check on awaiting processing" gap in the project managment interface

File: orchestra/project_api/serializers.py [hunk 7]
```diff
@@ -140,12 +141,10 @@ def get_worker(self, obj):
...
 
-    def get_snapshots(self, obj):
-        """
-        This function exists to automatically deserialize the JSON blob from
-        the `snapshots` JSONField
-        """
-        return obj.snapshots
+    def get_iterations(self, obj):
+        iterations = IterationSerializer(
+            obj.iterations.all().order_by('start_datetime'), many=True)
```
Comment (marcua) [inline orchestra/project_api/serializers.py]: no `all()`

File: orchestra/models.py [hunk 8]
```diff
@@ -435,16 +426,6 @@ class Status:
...
 
-    # When a worker submits, accepts, or rejects a task, we snapshot their
-    # in_workflow_task_data along with the date in the following format:
-    # {'snapshots': [
-    #   {'data': snapshotted_task_data,
-    #    'datetime': ISO 8601 datetime in UTC time,
-    #    'work_time_seconds': integer seconds,
-    #    'type': value from SnapshotType}]
-    #  '__version': 1}
-    snapshots = JSONField(default={}, blank=True)
```
Comment (lydiagu) [inline orchestra/models.py:446]: create a migration for this?

File: orchestra/static/orchestra/project_management/js/controllers.js [hunk 9]
```diff
@@ -6,8 +6,9 @@
     .controller('ProjectManagementController', ProjectManagementController);
 
   function ProjectManagementController($location, $scope, $routeParams, $http, $sce,
-    $compile, $modal, $timeout, projectVis) {
+    $compile, $modal, $timeout, projectVis, dataService) {
```
Comment (marcua) [inline orchestra/static/orchestra/project_management/js/controllers.js]: remove

File: orchestra/tests/test_project_management_api.py [hunk 10]
```diff
@@ -384,78 +329,56 @@ def test_end_project_api(self):
         for task in Task.objects.filter(project=project):
             self.assertEquals(task.status, Task.Status.ABORTED)
 
-    def _test_audit(self, audit, num_assignments,
```
Comment (marcua) [inline orchestra/tests/test_project_management_api.py:387]: put it back :)

File: orchestra/utils/revert.py [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+            An audit trail for the revert, e.g.,
+            {
+                'task': {...},
+                'change': 1
+                'assignments': [
+                    {
+                        # worker_0 assignment
+                        'assignment': {...},
+                        'change': 1,
+                        'iterations': [
```
Comment (marcua) [inline orchestra/utils/revert.py:53]: oi?

File: orchestra/utils/revert.py [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+    if assignment_audit['change'] == RevertChange.DELETED.value:
+        assignment.delete()
+    elif assignment_audit['change'] == RevertChange.REVERTED.value:
+        for iteration_audit in assignment_audit['iterations']:
+            _revert_iteration_from_audit(iteration_audit)
+
+        assignment.refresh_from_db()
+        if get_latest_iteration(
+                assignment.id).status == Iteration.Status.PROCESSING:
+            assignment.status = TaskAssignment.Status.PROCESSING
```
Comment (marcua) [inline orchestra/utils/revert.py:105]: leave a TODO for when we archive data to set the latest data to the last submitted data

File: orchestra/utils/revert.py [hunk 13]
```diff
@@ -0,0 +1,258 @@
+from enum import Enum
```
Comment (marcua) [inline orchestra/utils/revert.py:1]: is coverage 100%?

File: docs/source/api.rst (hunk 1 above)
Comment (jrbotros) [inline docs/source/api.rst]: :white_check_mark:

File: docs/source/api.rst (hunk 2 above)
Comment (jrbotros) [inline docs/source/api.rst]: :white_check_mark:

File: orchestra/analytics/latency.py (hunk 3 above)
Comment (jrbotros) [inline orchestra/analytics/latency.py:1]: :grin:

File: orchestra/machine_tasks.py (hunk 4 above)
Comment (jrbotros) [inline orchestra/machine_tasks.py]: Yes, but will revert as discussed offline since it's used internally!

Comment (jrbotros) [inline orchestra/machine_tasks.py]: :white_check_mark:

File: docs/source/api.rst (hunk 5 above)
Comment (jrbotros) [inline docs/source/api.rst]: :white_check_mark:

File: orchestra/project_api/serializers.py (hunk 7 above)
Comment (jrbotros) [inline orchestra/project_api/serializers.py]: :white_check_mark:

File: orchestra/models.py (hunk 8 above)
Comment (jrbotros) [inline orchestra/models.py:446]: Oops, wasn't added!

Comment (jrbotros) [inline orchestra/models.py:446]: :white_check_mark:

File: orchestra/static/orchestra/project_management/js/controllers.js (hunk 9 above)
Comment (jrbotros) [inline orchestra/static/orchestra/project_management/js/controllers.js]: :white_check_mark:

File: orchestra/utils/revert.py (hunk 12 above)
Comment (jrbotros) [inline orchestra/utils/revert.py:105]: :white_check_mark:

File: orchestra/utils/revert.py (hunk 11 above)
Comment (jrbotros) [inline orchestra/utils/revert.py:53]: :white_check_mark:

File: orchestra/utils/revert.py (hunk 13 above)
Comment (jrbotros) [inline orchestra/utils/revert.py:1]: Yes! :white_check_mark:

**Pull request merged**